Add grayMatter option to gatsby-plugin-mdx

riywo opened the pull request:
## Description
This commit is adding a new option for `gray-matter` used inside `gatsby-plugin-mdx`. The main motivation of this PR is for adding `excerpt_separator` which is commonly used in blogs.

When you specify:
```
module.exports = {
  plugins: [
    {
      resolve: `gatsby-plugin-mdx`,
      options: {
        grayMatter: {
          excerpt_separator: "<<<my-separator>>>"
        }
      }
    }
  ]
};
```

`gray-matter` automatically pick content before the `excerpt_separator`. ~~However, it contains mdx resources, so we need to translate it. In the `excerpt` resolver, it runs `processMDX()` for the extracted excerpt string just like it does for without `excerpt_separator`, then concat them.~~ Any MDX syntax won't be transformed since `gray-matter` doesn't know anything about MDX.

## Related Issues
Fixes #16865 

Note: Previously I tried to implement `excerpt_separator` inside the resolver but it seemed too complex. #16820

ChristopherBiscardi commented on packages/gatsby-plugin-mdx/gatsby/source-nodes.test.js:
```diff
@@ -0,0 +1,178 @@
+const { graphql } = require(`gatsby/graphql`)
```
it seems like this test requires mocking a lot of code, making it prone to breakage and extra maintenance burden in the future. Have any ideas to simplify this?

riywo replied:
This test was mostly imported from `gatsby-transfomer-remark`. I'll try to find the way to test GraphQL query easier...

riywo commented on packages/gatsby-plugin-mdx/gatsby/source-nodes.js:
```diff
@@ -131,19 +146,17 @@ module.exports = (
         async resolve(mdxNode, { pruneLength }) {
           if (mdxNode.excerpt) {
-            return Promise.resolve(mdxNode.excerpt)
+            const mdxExcerptNode = await createMDXNode({
```
`gray-matter` can't interpret any `import` or `export` statement. So, if we use the string excerpt, it could contain unnecessary text.

I'm ok with it for my use case, though. I'll revert this.

riywo commented on packages/gatsby-plugin-mdx/gatsby/on-create-node.js:
```diff
@@ -53,11 +53,14 @@ module.exports = async (
   createNode(mdxNode)
   createParentChildLink({ parent: node, child: mdxNode })
 
+  if (!cache) return
```
Yeah, this was added to pass the unit test. I'll revert and try to modify the test itself.

riywo commented on packages/gatsby-plugin-mdx/utils/gen-mdx.js:
```diff
@@ -76,10 +76,16 @@ module.exports = async function genMDX(
   debug(`processing classic frontmatter`)
-  const { data, content: frontMatterCodeResult } = grayMatter(node.rawBody)
+  const grayMatterOptions =
+    typeof options === `undefined` ? {} : options.grayMatter || {}
```
I see. Will do that.

riywo commented on packages/gatsby-plugin-mdx/utils/gen-mdx.js:
```diff
@@ -76,10 +76,16 @@ module.exports = async function genMDX(
+    node.rawBody,
+    grayMatterOptions
+  )
+  const frontmatter = excerpt === `` ? data : { ...data, excerpt }
```
Good point. I'll fix it.

riywo commented:
> How do you see this PR answering the open questions in #16865?

- Does the excerpt stay in the content afterwards?
    - I believe so. I'll try to add such test.
- gray-matter supports a lot of options, some of which are literal functions. Is there a better way to support those since gatsby-config is serialized?
    - I have no idea so far.
- How does the excerpt interact with the gatsby-plugin-mdx excerpt resolver. Currently if there is an excerpt we'd prefer it over the resolver.
    - As I commented, it can work by just leaving an excerpt found by gray-matter but it could contain unnecessary `export` or `import` statements.
- What happens to react components in excerpts? currently we don't render them.
    - Same above. It returns as is. We could strip `export` and `import` using regex.
- We should probably test what already works (like frontmatter language inference) before attempting to add this functionality.
    - Added a test but it seems too complex. Maybe we need finer-grained tests, not E2E for GraphQL.

riywo commented:
@ChristopherBiscardi I've updated the PR.

- Introduced `utils/gray-matter` to wrap the normal `gray-matter` and make testing easier.
- Add snapshot tests for `utils/gray-matter`.
    - So far, they are only focusing on `excerpt` related options. We can add more tests for other options later.
- Note: Any MDX syntax won't be transformed e.g. bare `import`, `export` or `JSX` including shortcodes.

riywo commented:
BTW, I think rendering Markdown (not MDX) is at least required like `gatsby-transformer-remark`'s `PLAIN` format...

muescha commented on packages/gatsby-plugin-mdx/utils/gray-matter/index.js:
```diff
@@ -0,0 +1,11 @@
+const grayMatter = require(`gray-matter`)
+
+const stripExcerpt = (excerpt, _options) => excerpt.trim().replace(/\n+/g, ` `)
```
question: is stripping new lines per default right here or should it return 1:1 the excerpt and its up to the user to change it as needed?

how is it done in [packages/gatsby-transformer-remark](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-transformer-remark) since there is also an excerpt option there ?

dabrady replied:
The `gatsby-transformer-remark` plugin [apparently uses `gray-matter`](https://using-remark.gatsbyjs.org/excerpts/) under the hood to retrieve the excerpt.

But if I'm understanding [the actual implementation](https://github.com/gatsbyjs/gatsby/blob/69ecf6a534b2be20e3b1ffe0b694ea677b68bb66/packages/gatsby-transformer-remark/src/extend-node-type.js#L436-L439) correctly, the `gray-matter` excerpt is only used in one particular case: when the `format:MARKDOWN` option is specified in the GraphQL query _and_ the `excerpt_separator` option is configured; in all other cases, the excerpt parsing is hand-rolled.

A quick test on my own site shows the behavior of the different excerpt formatting modes when an `excerpt_separator` is specified and the excerpt is rendered in a `pre` tag (to preserve whitespace and prevent text-wrapping).

| `PLAIN` | `MARKDOWN` | `HTML` |
:--------:|:----------:|:-------:
<img width="1680" alt="Screen Shot 2020-01-05 at 21 21 06" src="https://user-images.githubusercontent.com/2475919/71785601-ca93b600-3001-11ea-8666-ddb5b39f6bbe.png"> | <img width="1680" alt="Screen Shot 2020-01-05 at 21 21 16" src="https://user-images.githubusercontent.com/2475919/71785602-ca93b600-3001-11ea-8b15-4d096a9e7d3c.png"> | <img width="1680" alt="Screen Shot 2020-01-05 at 21 21 24" src="https://user-images.githubusercontent.com/2475919/71785603-ca93b600-3001-11ea-9c18-2dd0a76162d1.png">

The results are interesting. It seems `gatsby-transformer-remark` plugin is trying to preserve newlines (though in the case of `PLAIN` I think there is a slight bug: entirely blank lines cause the preceding and following line to be joined).

So if we are trying to mirror the behavior of the `Remark` plugin, I think we should preserve the output of `gray-matter` exactly, and let the user decide if they want to manipulate it further.

riywo replied:
Updated.

riywo commented:
With this change, it behaves apparently differently between with and without `excerpt_separator`: https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-plugin-mdx/gatsby/source-nodes.js#L133-L146

- If `grayMatter.excerpt_separato` is specified: The excerpt is not processed by MDX
- If not: The excerpt is processed by MDX

muescha commented on packages/gatsby-plugin-mdx/utils/mdx.js:
```diff
@@ -1,5 +1,5 @@
 const mdx = require(`@mdx-js/mdx`)
-const grayMatter = require(`gray-matter`)
+const grayMatter = require(`./gray-matter`)
```
I would name it different, otherwise other devs are expecting the original GrayMatter functionality here. 

one possible naming
```suggestion
const grayMatterWrapper = require(`./gray-matter-wrapper`)
```

riywo replied:
Thank you for the suggestion. I've updated.

pvdz commented on packages/gatsby-plugin-mdx/utils/gray-matter-wrapper/index.js:
```diff
@@ -0,0 +1,8 @@
+module.exports = (input, options) => {
+  const matter = grayMatter(input, options.grayMatter || {})
+  const { data: origData, excerpt, content } = matter
+  const data = excerpt === `` ? origData : { excerpt, ...origData }
```
Can `excerpt` not be `null` or `undefined` here in case of absence? Just wondering whether a `!excerpt` wouldn't be better here, if it can also be the empty string.

mathieudutour commented:
on `gatsby-transformer-remark`, the `excerpt_separator` option is at the root: https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-transformer-remark#example-excerpts

Shall it be the same here to be consistent?

riywo commented:
> on `gatsby-transformer-remark`, the `excerpt_separator` option is at the root: https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-transformer-remark#example-excerpts
> 
> Shall it be the same here to be consistent?

That's my first attempt but rejected. https://github.com/gatsbyjs/gatsby/pull/16820

Anyway, even if this is merged, the graymatter provided excerpt is useless since it is a plain text and the customer needs to process it by MDX (I don't know how I can do that). So, in my use case, I gave up this approach and decided to specify excerpt in frontmatter explicitly (used unified to scrape the excerpt content in the existing MD files.)

Thus, I don't have any motivation to push this change. If it has any use case, please reopen this PR.
